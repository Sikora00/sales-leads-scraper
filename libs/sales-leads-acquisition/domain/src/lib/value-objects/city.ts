import { TinyTypeOf } from 'tiny-types';

const type = Symbol();

export class City extends TinyTypeOf<string>() {
  private readonly type: typeof type = type;
}
