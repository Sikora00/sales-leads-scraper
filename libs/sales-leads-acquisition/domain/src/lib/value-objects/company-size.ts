import { TinyTypeOf } from 'tiny-types';

const type = Symbol();

export class CompanySize extends TinyTypeOf<string>() {
  private readonly type: typeof type = type;
}
