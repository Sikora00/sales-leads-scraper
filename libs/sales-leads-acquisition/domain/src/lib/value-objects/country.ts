import { TinyTypeOf } from 'tiny-types';

const type = Symbol();

export class Country extends TinyTypeOf<string>() {
  private readonly type: typeof type = type;
}
