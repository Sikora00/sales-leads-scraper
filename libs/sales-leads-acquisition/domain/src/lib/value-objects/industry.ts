import { TinyTypeOf } from 'tiny-types';

const type = Symbol();

export class Industry extends TinyTypeOf<string>() {
  private readonly type: typeof type = type;
}
