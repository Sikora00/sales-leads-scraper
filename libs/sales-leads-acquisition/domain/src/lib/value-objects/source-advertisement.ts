import { Url } from '@sales-leads/shared/domain-technical';

const type = Symbol();

export class SourceAdvertisement {
  private readonly type: typeof type = type;

  constructor(public readonly url: Url, public readonly title: string) {}
}
