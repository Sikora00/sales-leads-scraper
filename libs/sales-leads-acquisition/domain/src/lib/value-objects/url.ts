const type = Symbol();

export class Url {
  private readonly type: typeof type = type;
  private readonly value: URL;

  constructor(value: string) {
    this.value = new URL(value);
  }
}
