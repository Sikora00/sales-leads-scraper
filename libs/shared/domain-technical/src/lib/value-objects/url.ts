import { TinyTypeOf } from 'tiny-types';

const type = Symbol();

export class Url extends TinyTypeOf<URL>() {
  private readonly type: typeof type = type;

  constructor(value: string) {
    super(new URL(value));
  }

  toString(): string {
    return this.value.toString();
  }
}
