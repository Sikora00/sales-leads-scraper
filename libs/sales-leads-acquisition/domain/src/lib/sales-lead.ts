import { AggregateRoot } from '@nestjs/cqrs';
import { Url } from '@sales-leads/shared/domain-technical';
import { Country } from './value-objects/country';
import { Industry } from './value-objects/industry';
import { CompanySize } from './value-objects/company-size';
import { SourceAdvertisement } from './value-objects/source-advertisement';
import { City } from './value-objects/city';
import { NewSalesLeadAcquiredEvent } from './events/new-sales-lead-acquired.event';

export class SalesLead extends AggregateRoot {
  private city: City;
  private companySize: CompanySize;
  private companyWebsite: Url;
  private country: Country;
  private industry: Industry;
  private name: string;
  private sourceAdvertisement: SourceAdvertisement;

  private constructor(data: SalesLeadData) {
    super();
    this.city = data.city;
    this.companySize = data.companySize;
    this.companyWebsite = data.companyWebsite;
    this.country = data.country;
    this.industry = data.industry;
    this.name = data.name;
    this.sourceAdvertisement = data.sourceAdvertisement;
  }

  static acquired(data: SalesLeadData): SalesLead {
    const salesLead = new SalesLead(data);
    salesLead.apply(new NewSalesLeadAcquiredEvent());
    return salesLead;
  }

  static restore(data: SalesLeadData): SalesLead {
    return new SalesLead(data);
  }

  getCity(): City {
    return this.city;
  }

  getCompanySize(): CompanySize {
    return this.companySize;
  }

  getCompanyWebsite(): Url {
    return this.companyWebsite;
  }

  getCountry(): Country {
    return this.country;
  }

  getIndustry(): Industry {
    return this.industry;
  }

  getName(): string {
    return this.name;
  }

  getSourceAdvertisement(): SourceAdvertisement {
    return this.sourceAdvertisement;
  }
}

export interface SalesLeadData {
  city: City;
  companySize: CompanySize;
  companyWebsite: Url;
  country: Country;
  industry: Industry;
  name: string;
  sourceAdvertisement: SourceAdvertisement;
}
