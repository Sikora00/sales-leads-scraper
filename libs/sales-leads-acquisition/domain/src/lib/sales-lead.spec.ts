import { SalesLead, SalesLeadData } from './sales-lead';
import { City } from './value-objects/city';
import { Url } from './value-objects/url';
import { CompanySize } from './value-objects/company-size';
import { Country } from './value-objects/country';
import { Industry } from './value-objects/industry';
import { SourceAdvertisement } from './value-objects/source-advertisement';

describe('SalesLead', () => {
  describe('when sales lead is acquired', () => {
    let salesLead: SalesLead;
    beforeAll(() => {
      salesLead = SalesLead.acquired(fixture.getSalesLeadData());
    });
    it('should know its name, website, country, industry, company size, source advertisement, advertisement title and city', () => {
      expect(salesLead).toMatchSnapshot();
    });
  });
});

const fixture = {
  getSalesLeadData: (): SalesLeadData => ({
    city: new City('Warsaw'),
    companyWebsite: new Url('https://valueadd.pl'),
    companySize: new CompanySize('50-100'),
    country: new Country('Poland'),
    industry: new Industry('Software Development'),
    name: 'ValueAdd',
    sourceAdvertisement: new SourceAdvertisement(
      new Url('https://just-join-it.com'),
      'NestJS Developer'
    ),
  }),
};
