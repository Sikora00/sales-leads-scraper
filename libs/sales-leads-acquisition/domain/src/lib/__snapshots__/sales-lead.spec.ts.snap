// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalesLead when sales lead is acquired should know its name, website, country, industry, company size, source advertisement, advertisement title and city 1`] = `
SalesLead {
  "city": Object {
    "type": undefined,
    "value": "Warsaw",
  },
  "companySize": Object {
    "type": undefined,
    "value": "50-100",
  },
  "companyWebsite": Object {
    "type": undefined,
    "value": "https://valueadd.pl/",
  },
  "country": Object {
    "type": undefined,
    "value": "Poland",
  },
  "industry": Object {
    "type": undefined,
    "value": "Software Development",
  },
  "name": "ValueAdd",
  "sourceAdvertisement": SourceAdvertisement {
    "title": "NestJS Developer",
    "type": Symbol(),
    "url": Object {
      "type": undefined,
      "value": "https://just-join-it.com/",
    },
  },
  Symbol(): false,
  Symbol(): Array [
    NewSalesLeadAcquiredEvent {},
  ],
}
`;
